src/userobjects/PKAGeneratorAlphaDecay.C

```diff
@@ -0,0 +1,189 @@
+      decay._decay_constants[l] = std::log(2.0) / (half_life * _time_conversion);
+      decay._alpha_energies[l] = energy;
+      decay._intensities[l] = intensity;
+    }
+
+    if (std::find(_zaids.begin(), _zaids.end(), zaid) != _zaids.end())
+    {
+      if (_decay_data_sets.find(zaid) != _decay_data_sets.end())
+        mooseError("Identical Z/A id was provided twice");
+      _decay_data_sets[zaid] = decay;
+    }
+  }
+  infile.close();
+
+  // complete the data by dealing with absent nuclides ==> stable
+  for (auto & z : _zaids)
+  {
+    if (_decay_data_sets.find(z) == _decay_data_sets.end())
+    {
+      DecayData decay = DecayData(1);
+      decay._decay_constants[0] = 0.0;
+      decay._alpha_energies[0] = 0.0;
+      decay._intensities[0] = 1.0;
+      _decay_data_sets[z] = decay;
+    }
+  }
+
+  // in debug mode we want some info on what is stored in _decay_data_sets
+  #if DEBUG
+    _console << "\nAlpha decay read from file\n";
+    for (auto & d : _decay_data_sets)
+    {
+      _console << "ZAID: " << d.first << " has " << d.second._n_decay_channels << " decay channels:\n";
+      for (unsigned int j = 0; j < d.second._n_decay_channels; ++j)
+        _console << "lambda = " << d.second._decay_constants[j] << " "
+                 << "energy = " << d.second._alpha_energies[j] << " "
+                 << "intensity = " << d.second._intensities[j] << "\n";
+    }
+    _console << std::endl;
+  #endif
+}
+
+void
+PKAGeneratorAlphaDecay::appendPKAs(std::vector<MyTRIM_NS::IonBase> & ion_list, Real dt, Real vol, const MyTRIMRasterizer::AveragedData & averaged_data) const
+{
+  mooseAssert(dt >= 0, "Passed a negative time window into PKAFissionFragmentNeutronics::appendPKAs");
+  mooseAssert(vol >= 0, "Passed a negative volume into PKAFissionFragmentNeutronics::appendPKAs");
+
+  if (averaged_data._elements.size() != _zaids.size())
+    mooseError("Size of averaged_data and ZAID must be equal");
+
+  for (unsigned int nuclide = 0; nuclide < _zaids.size(); ++nuclide)
+  {
+    unsigned int zaid = _zaids[nuclide];
+
+    auto it = _decay_data_sets.find(zaid);
+
+    if (it == _decay_data_sets.end())
+      mooseError("ZAID ", zaid, " not found in decay data set");
+
+    auto & decay = it->second;
+    for (unsigned int l = 0; l < decay._n_decay_channels; ++l)
+    {
+      unsigned int num_decay = std::floor(vol * decay._intensities[l] * averaged_data._elements[nuclide]
+                                              * (1.0 - std::exp(-decay._decay_constants[l] * dt)) + getRandomReal());
+
+      for (unsigned i = 0; i < num_decay; ++i)
+      {
+        std::vector<MyTRIM_NS::IonBase> ion(2);
+
+        // set stopping criteria
+        ion[0].setEf();
+        ion[1].setEf();
+
+        // the tag is the element this PKA get registered as upon stopping
+        // -1 means the PKA will be ignored
+        ion[0]._tag = -1;
+        ion[1]._tag = -1;
+        if (_track_alpha)
+          ion[0]._tag = _alpha_index;
```

Comment:
Is this going to be superseded by the PR I just merged?

Reply:
Yes.

Reply:
Do you want to rebase this branch and make that change?

Reply:
Doing that right now.